### PR TITLE
Full GPIO Functionality

### PR DESCRIPTION
**Pull request by nullket**
**What this PR does**
This PR adds the ability to *optionally* set the two GPIOs of the cameras. If not set, it defaults to `input` in order to no break something.

**What can the GPIOs be set to?**
- Input
- Output Low
- Output High
- PWM (including frequency and duty cycle)
- Trigger (input)

**What else**
I have added an additional launch file to show an alternative way to create master-slaves setup. In this setup one camera creates a pwm output to trigger itself and other cameras

@jmackay2 again, I will ask for your time to review this please. I can imagine, that not everybody has the right cable. Therefore, you might can just review the code and see if it compiles fine in your setup without testing the GPIO functionality in case you do not have the right cable?

### CODE REVIEW

File: launch/master_slaves_pwm_rgb8.launch [hunk 1]
```diff
@@ -0,0 +1,167 @@
+<!-- In this example launch file one camera creates a pwm signal on GPIO 2 which can be used to
```
Comment (jmackay2) [inline launch/master_slaves_pwm_rgb8.launch:1]: Did you test this example? I am pretty sure the `<?xml version="1.0"?>` needs to be the first line, so probably move that up.

File: launch/master_slaves_pwm_rgb8.launch [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+     <param name="gpio2" type="int" value="4" /> <!-- 0 Input, 1 Output Low, 2 Output High, 3 Flash, 4 PWM, 5 Trigger input  -->
+     <param name="pwm_freq" type="double" value="1 " /> <!-- in hz, This will determine the framerate of all cameras -->
+     <param name="gpio1" type="int" value="5" /> <!-- 0 Input, 1 Output Low, 2 Output High, 3 Flash, 4 PWM, 5 Trigger input  -->
+
+     <param name="pixel_clock" type="int" value="30" />
+     </node>
+
+     <node pkg="nodelet" type="nodelet" name="ueye_cam_nodelet_slave1"
+          args="load ueye_cam/ueye_cam_nodelet nodelet_manager">
+     <param name="camera_name" type="str" value="master" /> <!-- == namespace for topics and services -->
```
Comment (jmackay2) [inline launch/master_slaves_pwm_rgb8.launch:68]: Since the camera name and topic are the same, ROS is going to throw an error about trying to advertise the same service multiple times.

File: launch/master_slaves_pwm_rgb8.launch [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+     <param name="camera_topic" type="str" value="image_raw" />
+     <param name="camera_id" type="int" value="0" /> <!-- 0 = any camera; 1+: camera ID -->
```
Comment (jmackay2) [inline launch/master_slaves_pwm_rgb8.launch:70]: Shouldn't we be incrementing the `camera_id` here?

File: launch/master_slaves_pwm_rgb8.launch [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+
+     
+     <param name="gpio1" type="int" value="5" /> <!-- 0 Input, 1 Output Low, 2 Output High, 3 Flash, 4 PWM, 5 Trigger input  -->
+
+     <param name="pixel_clock" type="int" value="30" />
+     </node>
+
+     <node pkg="nodelet" type="nodelet" name="ueye_cam_nodelet_slave2"
+          args="load ueye_cam/ueye_cam_nodelet nodelet_manager">
+     <param name="camera_name" type="str" value="master" /> <!-- == namespace for topics and services -->
```
Comment (jmackay2) [inline launch/master_slaves_pwm_rgb8.launch:119]: Same issue with the camera name and topic being the same.

File: launch/master_slaves_pwm_rgb8.launch [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+     <param name="camera_topic" type="str" value="image_raw" />
+     <param name="camera_id" type="int" value="0" /> <!-- 0 = any camera; 1+: camera ID -->
```
Comment (jmackay2) [inline launch/master_slaves_pwm_rgb8.launch:121]: Same comment about incrementing the camera_id.

**Review by jmackay2 — COMMENTED** [review]
Sorry it took me so long to get around to this. I don't have the cables around to test this, but I think the code looks fine. My main concern here is the new launch file, which seems like there could be some copy-paste errors. If you want to make sure you can run that file and test that it works as expected, that would be nice.

**Comment by nullket** [comment]
> Sorry it took me so long to get around to this. I don't have the cables around to test this, but I think the code looks fine. My main concern here is the new launch file, which seems like there could be some copy-paste errors. If you want to make sure you can run that file and test that it works as expected, that would be nice.

I am using a **similar** launch file like this in our application - which works perfectly fine. In order to be more compatible with most cameras I used my working and tested launch file, removed some settings, changed the naming to be close to the existing sync file and added the larger comment at the top. As I have multiple (mixed synced and non-synced) cams connected I did not test the launch file again. Thanks a lot for the very precise review!! I corrected the launch file and hardcoded my camera_ids (so I could test) - but this is the same with already existing master/slave [launch file](https://github.com/anqixu/ueye_cam/blob/master/launch/master_slaves_rgb8.launch)

**tl;dr:** I just fixed the file. Thanks for the review. Please merge the PR when you think it is fine now.

**Review by jmackay2 — APPROVED** [review]
